crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+    }
+}
+
+enum DefinitionType {
+    Union { ty: UnionType },
+    Type(Type),
+}
+
+impl DefinitionType {
+    fn into_type(self, types: &mut TypeInferenceBuilder) -> Type {
+        match self {
+            DefinitionType::Union { ty } => types.union_ty(ty),
+            DefinitionType::Type(ty) => ty,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ruff_db::file_system::FileSystemPathBuf;
+    use ruff_db::vfs::system_path_to_file;
+
+    use crate::db::tests::TestDb;
+    use crate::module::resolver::{set_module_resolution_settings, ModuleResolutionSettings};
+    use crate::name::Name;
+    use crate::red_knot::types::{public_symbol_ty_by_name, Type, TypingContext};
+
+    fn setup_db() -> TestDb {
+        let mut db = TestDb::new();
+
+        set_module_resolution_settings(
+            &mut db,
+            ModuleResolutionSettings {
+                extra_paths: Vec::new(),
+                workspace_root: FileSystemPathBuf::from("/src"),
+                site_packages: None,
+                custom_typeshed: None,
+            },
+        );
+
+        db
+    }
+
+    fn assert_public_ty(db: &TestDb, file_name: &str, symbol_name: &str, expected: &str) {
+        let file = system_path_to_file(db, file_name).expect("Expected file to exist.");
+
+        let ty = public_symbol_ty_by_name(db, file, symbol_name).unwrap_or(Type::Unknown);
+        assert_eq!(ty.display(&TypingContext::global(db)).to_string(), expected);
+    }
+
+    #[test]
+    fn follow_import_to_class() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_files([
+            ("src/a.py", "from b import C as D; E = D"),
+            ("src/b.py", "class C: pass"),
+        ])?;
+
+        assert_public_ty(&db, "src/a.py", "E", "Literal[C]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn resolve_base_class_by_name() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_file(
+            "src/mod.py",
+            r#"
+class Base:
+    pass
+
+class Sub(Base):
+    pass"#,
+        )?;
+
+        let mod_file = system_path_to_file(&db, "src/mod.py").expect("Expected file to exist.");
+        let ty = public_symbol_ty_by_name(&db, mod_file, "Sub").expect("Symbol type to exist");
+
+        let Type::Class(class_id) = ty else {
+            panic!("Sub is not a Class")
+        };
+
+        let context = TypingContext::global(&db);
+
+        let base_names: Vec<_> = class_id
+            .lookup(&context)
+            .bases()
+            .iter()
+            .map(|base_ty| format!("{}", base_ty.display(&context)))
+            .collect();
+
+        assert_eq!(base_names, vec!["Literal[Base]"]);
+
+        Ok(())
+    }
+
+    #[test]
+    fn resolve_method() -> anyhow::Result<()> {
+        let db = setup_db();
+
+        db.memory_file_system().write_file(
+            "src/mod.py",
+            "
+                class C:
+                    def f(self): pass
```

Comment:
I'm not sure how this test (and others below) are passing, since we aren't using `dedent` on this string. Is our parser more forgiving about extra common leading whitespace than CPython's parser?

Reply:
Yeah, our parser fully recovers from parse errors and we aren't querying the parse errors in these tests.